lib/miniproxy.rb

```diff
@@ -50,20 +51,26 @@ def self.allow_request(method:, url:)
   end
 
   private_class_method def self.remote
-    Timeout.timeout(self.drb_timeout) do
-      begin
-        remote = DRbObject.new(nil, Remote.server(self.host))
+    started_at = GET_TIME.call(Process::CLOCK_MONOTONIC)
 
-        until remote.started?
-          sleep 0.01
-        end
+    begin
+      remote = DRbObject.new(nil, Remote.server(host))
 
-        remote.drain_messages.each(&method(:puts))
+      until remote.started?
+        sleep 0.01
 
-        remote
-      rescue
-        retry
+        if (GET_TIME.call(Process::CLOCK_MONOTONIC) - started_at) < drb_timeout
+          raise Timeout::Error
+        end
       end
+
+      remote.drain_messages.each(&method(:puts))
+
+      remote
+    rescue Timeout::Error
+      raise
+    rescue
+      retry
```

Comment:
There is a chance for an infinite loop here -  for example if `remote.drain_messages.each(&method(:puts))` always errored but the `remote.started?` is `true` the timeout won't be checked.
 
There might be a way to restructure this so it's safer- maybe we can assume that `remote.started?` is the only code that we want to rescue errors from. Alternately we could count the retires and only call `retry` if it's under a certain number.

Reply:
Good catch, I think it's fair to assume we only want to retry errors from `remote.started?`.

Reply:
I like the update 👍